test/test_position_endpoint.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+import pytest
+
+from girder.models.item import Item
+from girder.models.folder import Folder
+from pytest_girder.assertions import assertStatus, assertStatusOk
+
+
+@pytest.fixture
+def makeResources(admin):
+    publicFolder = Folder().findOne({'parentId': admin['_id'], 'name': 'Public'})
+    folders = []
+    publicItems = []
+    privateItems = []
+    for i in range(20):
+        folder = Folder().createFolder(
+            parent=publicFolder, creator=admin, parentType='folder', name='Folder %d' % i)
+        if i % 2:
+            folder = Folder().setPublic(folder, False, save=True)
+        folders.append(folder)
+    for i in range(20):
+        publicItems.append(Item().createItem(
+            name='Public Item %d' % i, creator=admin, folder=folders[0]))
+        privateItems.append(Item().createItem(
+            name='Private Item %d' % i, creator=admin, folder=folders[1]))
+    yield {
+        'publicFolder': publicFolder,
+        'folders': folders,
+        'publicItems': publicItems,
+        'privateItems': privateItems,
+    }
+
+
+def testFolderPosition(server, makeResources, admin, user):
+    resp = server.request(
+        '/folder/%s/position' % str(makeResources['folders'][10]['_id']),
+        user=admin,
+        params={
+            'parentType': 'folder',
+            'parentId': str(makeResources['publicFolder']['_id'])
+        })
+    assertStatusOk(resp)
+    # alphabetical by C sort
+    assert resp.json == 2
+
+    resp = server.request(
```

Comment:
I think the best practice in pytest is to have single functions per case, where each has the elements of:
* initialization
* execution
* assertion

This has the benefit of allowing all cases to run, even if some fail; with a single function + multiple case approach, the first failed case prevents all subsequent cases from executing. It also has the benefit of ensuring that the system state is reset between each case.

Ideally, I hope these can each look like:
```python
def test_case(prereq_1, prereq_2, baseline_1):
    result = func_to_test(prereq_1, prereq_2)
    assert result.prop == baseline_1
```

https://github.com/girder/girder/blob/master/test/test_access.py has some good examples of this pattern.

Reply:
I'm going to disagree on your assertion of best practices.

Tests exist to serve development, not to be paragons of coding on their own.  Part of serving development is to be efficient.  There is a cognitive tradeoff in understanding tests that are many little pieces, all nearly alike, and having monolithic tests with state.  Neither is particularly easy to maintain -- if you have many small tests and there are problems, you have to update many things; if you have one test with many sub sections, it is harder to understand what the whole is doing and you risk having state that is implied in some way.

Further, many small tests each with do their own initialization tend to be slower than a single test with one initialization with a series of executions and assertions.  Slow tests tend not to get run as often as they add friction to development.

We expect all tests to pass, so the granular failure is of no particular utility; in practice when tests fail you fix the issues that cause them, and, if tests are quick to run, run them locally and iterate.  If they are slow to run, they get kicked to CI for running.

Having said all that, I have no problem changing these tests, but I don't think it actually improves anything (though it does make them run slower).

Reply:
@mgrauer recently linked [this article](https://www.arp242.net/testing.html) which I think everybody should read. I like a lot of what it says about testing. It's important for us to engage in debate about testing best practices. My main desire is that when we debate these things, that no one advocates anything in pursuit of dogma; we should always frame the debate in terms of practical value. (I'm not accusing anyone is being dogmatic here, I just want to set that out as a ground rule I hope we can all agree on.)

Anyway, one thing that really stuck out to me about that article is that the author's primary value metric in testing is *debuggability*, and I think that would also be a really good thing for us to focus on, because IMO debugging is where a lot of our costs associated with testing end up. If one were to argue *against* "larger" test cases, I think a convincing argument should involve debuggability.

Isolation of test cases is critically important for debuggability and reduction of flakiness. Far in the past I worked on a codebase at the wrong extreme of this continuum, where every single test ran against the same database, and state was not reset in between each one. As I'm sure you can imagine, it was an absolute nightmare to debug; tests had to be run in order, adding a new case to an existing test was disastrous since it often broke all the downstream ones. The tests become totally unmaintained as a result.

We're just trying to find the right balance between that nightmare, and on the other extreme, the misguided land of unit-test dogma. My personal take is that we should typically strive for @brianhelba 's style of test, but we shouldn't slavishly adhere to it. That is, when we want to write a new test, we should *try* to write one with that pattern, but if that ends up being annoying or not a good fit, we should be OK to write a larger test case with more sequential state changes. We don't need to bang our head against the wall trying to make it look a certain way if that just doesn't fit with the feature we need to test.